src/iconforge.rs

```diff
@@ -0,0 +1,1088 @@
+        error: error.lock().unwrap().join("\n"),
+    };
+    Ok(serde_json::to_string::<SpritesheetResult>(&returned)?)
+}
+
+/// Given an array of 'transform arrays' onto from a shared IconObject base,
+/// recursively applies transforms in a tree structure. Maximum transform depth is 128.
+fn transform_leaves(icons: &Vec<&IconObject>, image: RgbaImage, depth: u8) -> Result<(), String> {
+    zone!("transform_leaf");
+    if depth > 128 {
+        return Err(String::from(
+            "Transform depth exceeded 128. https://www.youtube.com/watch?v=CUjrySBwi5Q",
+        ));
+    }
+    let next_transforms = DashMap::<Transform, Vec<&IconObject>>::new();
+    let errors = Mutex::new(Vec::<String>::new());
+
+    {
+        zone!("get_next_transforms");
+        icons.par_iter().for_each(|icon| {
+            zone!("collect_icon_transforms");
+            if let Some(transform) = icon.transform.get(depth as usize) {
+                next_transforms
+                    .entry(transform.clone())
+                    .or_default()
+                    .push(icon);
+            }
+        });
+    }
+
+    {
+        zone!("do_next_transforms");
+        next_transforms
+            .into_par_iter()
+            .for_each(|(transform, mut associated_icons)| {
+                let mut altered_image;
+                {
+                    zone!("clone_image");
+                    altered_image = image.clone();
+                }
+                if let Err(err) = transform_image(&mut altered_image, &transform) {
+                    errors.lock().unwrap().push(err);
+                }
+                {
+                    zone!("filter_associated_icons");
+                    associated_icons
+                        .clone()
+                        .into_iter()
+                        .enumerate()
+                        .for_each(|(idx, icon)| {
+                            if icon.transform.len() as u8 == depth + 1
+                                && *icon.transform.last().unwrap() == transform
+                            {
+                                associated_icons.swap_remove(idx);
+                                if let Err(err) = return_image(altered_image.clone(), icon) {
+                                    errors.lock().unwrap().push(err.to_string());
+                                }
+                            }
+                        });
+                }
+                if let Err(err) = transform_leaves(&associated_icons, altered_image, depth + 1) {
+                    errors.lock().unwrap().push(err);
+                }
+            });
+    }
+
+    if !errors.lock().unwrap().is_empty() {
+        return Err(errors.lock().unwrap().join("\n"));
+    }
+    Ok(())
+}
+
+/// Converts an IO icon to one with icon_hash_input
+fn icon_from_io(icon_in: IconObjectIO) -> IconObject {
+    zone!("icon_from_io");
+    // TODO: can probably convert this function to just lazily attaching icostring to a RefCell<> or something
```

Comment:
todo?

Reply:
Yeah I'm not really sure how to do this and it works, it's just a potential microop for memory and processing to clone stuff